cpp/src/neighbors/detail/cagra/cagra_search.cuh

```diff
@@ -108,6 +111,109 @@ void search_main_core(raft::resources const& res,
+ * @param[out] neighbors a device matrix view to the indices of the neighbors in the source dataset
+ * [n_queries, k]
+ * @param[out] distances a device matrix view to the distances to the selected neighbors [n_queries,
+ * k]
+ * @param[in] sample_filter a device filter function that greenlights samples for a given query
+ * @param[in] threshold_to_bf A sparsity threshold; brute force is used when sparsity exceeds this
+ * threshold, in the range [0, 1]
+ *
+ * @return true If the brute force search was applied successfully.
+ * @return false If the brute force search was not applied.
+ */
+template <typename T,
+          typename InternalIdxT,
+          typename CagraSampleFilterT,
+          typename IdxT      = uint32_t,
+          typename DistanceT = float>
+bool search_using_brute_force(
+  raft::resources const& res,
+  search_params& params,
+  const index<T, IdxT>& index,
+  raft::device_matrix_view<const T, int64_t, raft::row_major>& queries,
+  raft::device_matrix_view<InternalIdxT, int64_t, raft::row_major>& neighbors,
+  raft::device_matrix_view<DistanceT, int64_t, raft::row_major>& distances,
+  CagraSampleFilterT& sample_filter,
+  double threshold_to_bf = 0.9)
+{
+  bool is_applied = false;
+  auto n_queries  = queries.extent(0);
+  auto n_dataset  = index.size();
+
+  auto bitset_filter_view = sample_filter.bitset_view_;
+  auto dataset_view       = index.contiguous_dataset();
+  auto sparsity           = bitset_filter_view.sparsity(res);
+
+  // TODO: Support host dataset in `brute_force::build`
+  if (sparsity >= threshold_to_bf &&
+      std::holds_alternative<raft::device_matrix_view<const T, int64_t, raft::row_major>>(
+        dataset_view)) {
+    using bitmap_view_t = cuvs::core::bitmap_view<const uint32_t, int64_t>;
+
+    auto stream            = raft::resource::get_cuda_stream(res);
+    auto bitmap_n_elements = bitmap_view_t::eval_n_elements(bitset_filter_view.size() * n_queries);
+
+    rmm::device_uvector<uint32_t> raw_bitmap(bitmap_n_elements, stream);
+    rmm::device_uvector<int64_t> raw_neighbors(neighbors.size(), stream);
+
+    bitset_filter_view.repeat(res, n_queries, raw_bitmap.data());
+
+    auto brute_force_filter = bitmap_view_t(raw_bitmap.data(), n_queries, n_dataset);
+
+    auto brute_force_neighbors = raft::make_device_matrix_view<int64_t, int64_t, raft::row_major>(
+      raw_neighbors.data(), neighbors.extent(0), neighbors.extent(1));
+    auto brute_force_dataset =
+      std::get_if<raft::device_matrix_view<const T, int64_t, raft::row_major>>(&dataset_view);
+
+    if (brute_force_dataset) {
+      RAFT_LOG_DEBUG("CAGRA is switching to brute force with sparsity:%d", sparsity);
+      auto brute_force_idx =
+        cuvs::neighbors::brute_force::build(res, *brute_force_dataset, index.metric());
+
+      auto brute_force_queries = queries;
+      auto padding_queries     = raft::make_device_matrix<T, int64_t>(res, 0, 0);
```

Comment:
I'd prefer to stick to raft primitives and eventually remove `copy_with_padding`, because the latter is not a commonly used primitive and thus one requires to go read its code to understand what does it do exactly.
But I don't mind to keep it here for now.